cpp/tests/utilities_tests/device_atomics_test.cu

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <tests/utilities/column_wrapper.cuh>
+#include <tests/utilities/cudf_test_fixtures.h>
+#include <utilities/wrapper_types.hpp>
+#include <utilities/device_atomics.cuh>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <thrust/device_vector.h>
+#include <thrust/transform.h>
+
+#include <bitset>
+#include <cstdint>
+#include <random>
+
+template<typename T>
+__global__
+void gpu_atomic_test(T *result, T *data, size_t size)
+{
+    size_t id   = blockIdx.x * blockDim.x + threadIdx.x;
+    size_t step = blockDim.x * gridDim.x;
+
+    for (; id < size; id += step) {
+        atomicAdd(&result[0], data[id]);
+        atomicMin(&result[1], data[id]);
+        atomicMax(&result[2], data[id]);
+        atomicAdd(&result[3], data[id]);
+    }
+}
+
+template<typename T>
+__global__
+void gpu_atomicCAS_test(T *result, T *data, size_t size)
+{
+    size_t id   = blockIdx.x * blockDim.x + threadIdx.x;
+    size_t step = blockDim.x * gridDim.x;
+
+    for (; id < size; id += step) {
+        T* addr = &result[0];
+        T update_value = data[id];
+
+        T old_value = *addr;
+        T assumed;
+
+        do {
+            assumed  = old_value;
+            const T new_value = old_value + update_value;
+
+            old_value = atomicCAS(addr, assumed, new_value);
+        } while (assumed != old_value);
+    }
+}
+
+// TODO: remove these explicit instantiation for kernels
```

Comment:
Why is it TODO to remove them? I think they are required.  You could delete this TODO.

Reply:
I think the explicit instantiation is a workaround, and it is compiler's issue related to google test's macro.
Actually I could reproduce the issue independent from cudf lib.
If compiler fixes the issue, we may remove the explicit instantiation.

Reply:
I see. Thanks.